File: Linear Shoal/Monotonic Stack/Q2. Daily Temperatures/Q2_DailyTemperatures.java
import java.util.*;

class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int n = temperatures.length;
        int[] answer = new int[n];         // Result array to store number of days to wait
        Stack<Integer> stack = new Stack<>(); // Stack to store indices of temperatures

        // Traverse the temperature array from end to start
        for (int i = n - 1; i >= 0; i--) {
            // Pop all indices with temperature less than or equal to current day's temperature
            while (!stack.isEmpty() && temperatures[i] >= temperatures[stack.peek()]) {
                stack.pop();
            }

            // If stack is not empty, calculate the days to wait
            if (!stack.isEmpty()) {
                answer[i] = stack.peek() - i;
            }

            // Push current index onto the stack
            stack.push(i);
        }

        return answer;
    }
}

File: Linear Shoal/Monotonic Stack/Q2. Daily Temperatures/Q2_DailyTemperatures.py
class Solution(object):
    def dailyTemperatures(self, temperatures):
        """
        :type temperatures: List[int]
        :rtype: List[int]
        """
        n = len(temperatures)              # Total number of days
        answer = [0] * n                   # Initialize answer list with 0s
        stack = []                         # Stack to store indices

        # Iterate from the end of the list to the beginning
        for i in range(n - 1, -1, -1):
            current_temp = temperatures[i]

            # Pop elements from the stack while current temp >= temp at top of stack
            while stack and temperatures[stack[-1]] <= current_temp:
                stack.pop()

            # If stack is not empty, top of stack is the next warmer day
            if stack:
                answer[i] = stack[-1] - i  # Calculate the number of days to wait

            # Push current index onto the stack
            stack.append(i)

        return answer
